src/inventory-service/inventory-core/src/main/java/com/inventory/core/InventoryItemService.java

```diff
@@ -71,95 +85,201 @@ public HandlerResponse<InventoryItemDTO> updateStock(UpdateInventoryStockRequest
-        } catch (Error error) {
-            logger.error("An exception occurred!", error);
-            span.setTag(Tags.ERROR, true);
-            span.log(Collections.singletonMap(Fields.ERROR_OBJECT, error));
-
-            return new HandlerResponse<>(null, List.of("Unknown error"), false);
+        } catch (Exception e) {
+            logger.error("Error updating stock", e);
+            if (span != null) {
+                span.setTag(Tags.ERROR, true);
+                span.log(Collections.singletonMap(Fields.ERROR_OBJECT, e));
+            }
+            return new HandlerResponse<>(null, List.of("Unknown error: " + e.getMessage()), false);
         }
     }
 
     public HandlerResponse<Boolean> reserveStockForOrder(String orderNumber, List<String> products, String conversationId) {
         final Span span = GlobalTracer.get().activeSpan();
+        logger.info("Reserving stock for order {} with {} products", orderNumber, products.size());
 
-        try {
-            var isFailure = false;
-            ArrayList<InventoryItem> stockAddedFor = new java.util.ArrayList<>();
-
+        if (span != null) {
             span.setTag("order.id", orderNumber);
             span.setTag("order.productCount", products.size());
+            span.setTag("order.conversationId", conversationId);
+        }
 
-            for (var productId : products) {
-                final Span stockCheckSpan = GlobalTracer.get()
-                        .buildSpan("stockCheck")
-                        .asChildOf(span)
-                        .start();
-
-                try (Scope scope = GlobalTracer.get().activateSpan(stockCheckSpan)) {
-                    stockCheckSpan.setTag("product.id", productId);
-                    var inventoryItem = this.repository.withProductId(productId);
+        // Store the products in the order cache early to ensure they're recorded even if reservation fails
+        try {
+            orderCache.store(orderNumber, new ArrayList<>(products));
+        } catch (Exception e) {
+            logger.error("Failed to store order in cache", e);
+            if (span != null) {
+                span.setTag("error.orderCache", true);
+                span.setTag("error.message", e.getMessage());
+            }
+            // Continue despite cache failure - we'll try to process the order anyway
+        }
 
-                    if (inventoryItem.getAvailableStockLevel() <= 0) {
-                        stockCheckSpan.setTag("product.outOfStock", "true");
-                        isFailure = true;
-                        break;
+        try {
+            AtomicBoolean isFailure = new AtomicBoolean(false);
+            List<InventoryItem> stockAddedFor = Collections.synchronizedList(new ArrayList<>());
+
+            // Process products in parallel for large orders for better performance
+            if (products.size() > 5) {
```

Comment:
I don't know why you bother specialising this; it just adds complexity to the code for not much gain; do it always in parallel so you have less to think about